crates/red_knot_module_resolver/src/resolver.rs

```diff
@@ -152,57 +161,308 @@ impl RawModuleResolutionSettings {
+    dynamic_paths: Option<ordermap::set::Iter<'db, SearchPathRoot>>,
+}
+
+impl<'db> SearchPathIterator<'db> {
+    fn new(db: &'db dyn Db) -> Self {
+        let static_paths = module_resolver_settings(db)
+            .search_path_settings
+            .static_search_paths()
+            .into_iter();
+
+        Self {
+            db,
+            static_paths,
+            dynamic_paths: None,
+        }
+    }
+}
+
+impl<'db> Iterator for SearchPathIterator<'db> {
+    type Item = &'db SearchPathRoot;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let SearchPathIterator {
+            db,
+            static_paths,
+            dynamic_paths,
+        } = self;
+
+        static_paths.next().or_else(|| {
+            dynamic_paths
+                .get_or_insert_with(|| dynamic_module_resolution_paths(*db).into_iter())
+                .next()
+        })
+    }
+}
 
-impl Deref for OrderedSearchPaths {
-    type Target = [Arc<ModuleResolutionPathBuf>];
+impl<'db> FusedIterator for SearchPathIterator<'db> {}
 
-    fn deref(&self) -> &Self::Target {
-        &self.0
+fn iter_module_search_paths(db: &dyn Db) -> SearchPathIterator {
+    SearchPathIterator::new(db)
+}
+
+/// Represents a single `.pth` file in a `site-packages` directory.
+/// One or more lines in a `.pth` file may be a (relative or absolute)
+/// path that represents an editable installation of a package.
+struct PthFile<'db> {
+    db: &'db dyn Db,
+    path: SystemPathBuf,
+    contents: SourceText,
+    site_packages: &'db SystemPath,
+}
+
+impl<'db> PthFile<'db> {
+    fn new(
+        db: &'db dyn Db,
+        pth_path: SystemPathBuf,
+        file: File,
+        site_packages: &'db SystemPath,
+    ) -> Self {
+        Self {
+            db,
+            path: pth_path,
+            contents: source_text(db.upcast(), file),
+            site_packages,
+        }
+    }
+
+    /// Yield paths in this `.pth` file that appear to represent editable installations,
+    /// and should therefore be added as module-resolution search paths.
+    fn iter_editable_installations<'a>(
+        &'a self,
+    ) -> impl Iterator<Item = ModuleResolutionPathBuf> + 'db
+    where
+        'a: 'db,
+    {
+        // Empty lines or lines starting with '#' are ignored by the Python interpreter.
+        // Lines that start with "import " or "import\t" do not represent editable installs at all;
+        // instead, these are files that are executed by Python at startup.
+        // https://docs.python.org/3/library/site.html#module-site
+        self.contents.lines().filter_map(|line| {
+            let line = line.trim();
+            if line.is_empty()
+                || line.starts_with('#')
+                || line.starts_with("import ")
+                || line.starts_with("import\t")
+            {
+                return None;
+            }
+            let possible_editable_install = SystemPath::absolute(
+                self.site_packages.join(line),
```

Comment:
Is it adviced to trim the path?

Reply:
Not sure I understand the question here, sorry!

Reply:
I wonder what happens if you have a line like `a/b                                                          ` (a lot of space). Should this resolve to `a/b` or throw because that directory probably doesn't exist on disk :D

Reply:
Hehe. The Python spec doesn't seem to say anything about preceding and trailing whitespace on each line of a `.pth` file, and I don't know of any build backends that add preceding or trailing whitespace. Pyright trims preceding and trailing whitespace from each line, however (I assume because it wants to err on the side of leniency), and Charlie's port does too. It seems reasonable to me to trim trailing whitespace; I already do that in the `PthFile::editable_installations()` method

Reply:
Python at runtime only strips trailing whitespace, not leading whitespace: https://github.com/python/cpython/blob/main/Lib/site.py#L208

I also verified that behavior experimentally.

I think we should do the same thing the runtime does.